### PR TITLE
[Containers] Add Array::Map

### PR DESCRIPTION
**Pull request by Lunderberg**
Previously, an in-place mutation could be applied to an array using `Array::MutateByApply`, but this couldn't be used for transformations that return a new array, or for transformations that return a new type.  The PR adds `Array::Map`, which can map to any `ObjectRef` subclass.  For mappings that return the same type, this is done by delegating to `Array::MutateByApply`, to take advantage of the same copy-on-write behavior.

For ease of review, this PR consists of two separate commits.  The first implements `Array::Map`, while the second performs several small refactors

### CODE REVIEW

File: include/tvm/runtime/container/array.h [hunk 1]
```diff
@@ -574,6 +574,44 @@ class Array : public ObjectRef {
...
+   * \note This function performs copy on write optimization.  If
+   * `fmap` returns an object of type `T`, and all elements of the
+   * array are mapped to themselves, then the returned array will be
+   * the same as the original, and reference counts of the elements in
+   * the array will not be incremented.
+   *
+   * \return The transformed array.
+   */
+  template <typename F, typename U = std::invoke_result_t<F, T>>
+  Array<U> Map(F fmap) const {
```
Comment (junrushao) [inline include/tvm/runtime/container/array.h:597]: I was curious if we could unify and migrate `MutateByApply` into `Map`?

Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:597]: Possibly, and that would allow for avoiding copies in a few additional cases (e.g. map from `T` to `Optional<T>`, or to a superclass of `T`) that aren't currently handled.  I'll take a quick stab at it and see if I can unify the two.

Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:616]: Thank you on the suggestion, and it ended up being much cleaner that way.  Both `Map` and `MutateByApply` are now implemented in terms of the same underlying helper function.  The helper function applies both the mutate-in-place and copy-on-write optimizations, with `if constexpr` type checks to avoid attempting the optimization if they wouldn't be possible.

File: include/tvm/runtime/container/array.h [hunk 2]
```diff
@@ -706,6 +710,98 @@ class Array : public ObjectRef {
...
+    } else {
+      // Path for incompatible types.  The constexpr check for
+      // compatible types isn't strictly necessary, as the first
+      // mapped.same_as(*it) would return false, but we might as well
+      // avoid it altogether.
+      output = ArrayNode::CreateRepeated(arr->size(), U());
+    }
+
+    // Normal path for incompatible types, or post-copy path for
+    // copy-on-write instances.
```
Comment (janetsc) [inline include/tvm/runtime/container/array.h:797]: What will be left over on the copy-on-write instance?  Will there be some items that are incompatible?  How are those guaranteed to be at the end?

File: include/tvm/runtime/container/array.h [hunk 3]
```diff
@@ -706,6 +710,98 @@ class Array : public ObjectRef {
...
+    for (; it != arr->end(); it++) {
+      U mapped = fmap(DowncastNoCheck<T>(*it));
+      output->SetItem(it - arr->begin(), std::move(mapped));
+    }
+
+    return output;
+  }
```
Comment (janetsc) [inline include/tvm/runtime/container/array.h:804]: General comment - Can you add a unit test to exercise the edge cases in MapHelper?

File: include/tvm/runtime/container/array.h (hunk 2 above)
Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:797]: > What will be left over on the copy-on-write instance? 

If we have compatible types, and we've reached this point, we've found at least one element for which the `mapped.same_as(*it)` check on line 776 has failed.  In that case, `output` will contain everything in the range `[arr->begin(), it)`.  That is, `output` contains all elements that are identical, and the first non-identical element.  `it` will point to the next element that should be transformed, and so the next loop over `it` can continue where the first loop left off.

> Will there be some items that are incompatible?

It's entirely possible, either at compile-time or at runtime.  For example, I could have an `Array<PrimExpr> buffer_shape` and map it to allowed ranges `buffer_shape.Map([](PrimExpr expr) { return Range::FromMinExtent(0, expr);});`, which would be incompatible and identified as such at compile-time.  In that case, the `if constexpr` could identify that they cannot be represented by the same underlying array, and can skip the attempts to do so altogether.

If a type is incompatible at runtime, then it will also fail the `mapped.same_as(*it)` check on line 776.  So if I have an `Array<Var>` being mapped to `Array<PrimExpr>` with `var_array.Map([&](Var var) { return var.same_as(to_replace) ? replace_with : var;});`, it may or may not be compatible, depending on whether `to_replace` shows up in the array.

> How are those guaranteed to be at the end?

Incompatible items may occur at any point in the mapped output, even at the very first iteration.  In that case, the commands executed in the conditional on `!mapped.same_as(*it)` are the same as would be executed up through the first iteration of the mapping loop.

```c++
// Same as the else branch on `compatible_types`
output = ArrayNode::CreateRepeated(arr->size(), U());

// For the first iteration, it is `arr->begin()`, so this would be an
// empty range [begin, begin), nothing is initialized, and this
// statement has no effect.
output->InitRange(0, arr->begin(), it);

// The newly mapped item is stored to the first location of the output.
output->SetItem(it - arr->begin(), std::move(mapped));

// The loop increment that would have happened
it++;

// `it` now points to the second element of the input, and we have one
// mapped element in the output.  We're now ready to start the second
// loop, just at the second iteration instead of the first.
```

Essentially, we only need to check for identical return values up until we find a single non-identical element, at which point we know that we can't avoid the copy anyways.  But once we reach the first non-identical value, we don't need to repeat the function calls up to that point, because we know that everything is either identical (and can therefore be copied from the input) or is non-identical (is which case it is the first such non-identical value).

File: include/tvm/runtime/container/array.h (hunk 3 above)
Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:804]: Certainly, and thank you for pointing that out!  There are some existing tests in [`container_test.cc`](https://github.com/apache/tvm/blob/main/tests/cpp/container_test.cc#L165), along with a large amount of usage when lowering TIR, but no tests that would specifically point to these edge cases.

Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:804]: Tests added for each of the compatible types, to validate that copies are avoided, and to ensure correct fail-through behavior when a copy is required.  A double thanks for requesting it, as it also caught a type conversion error that I had missed.

File: include/tvm/runtime/container/array.h (hunk 2 above)
Comment (janetsc) [reply include/tvm/runtime/container/array.h:797]: Thanks for this explanation!

**Review by janetsc — APPROVED** [review]
Unit test looks good - thanks!

File: include/tvm/runtime/container/array.h (hunk 2 above)
Comment (Lunderberg) [reply include/tvm/runtime/container/array.h:802]: No problem, and updated!

**Review by junrushao — APPROVED** [review]
LGTM!

**Comment by tmoreau89** [comment]
Awesome thank you @Lunderberg @janetsc @junrushao , the PR has been merged!
